source/commoncrypto/commoncrypto_ecc.c

```diff
@@ -0,0 +1,607 @@
+        (priv_key && priv_key->len != s_key_coordinate_size)) {
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        goto error;
+    }
+
+    size_t total_buffer_size = s_key_coordinate_size * 3 + 1;
+
+    if (aws_byte_buf_init(&cc_key_pair->key_pair.key_buf, allocator, total_buffer_size)) {
+        goto error;
+    }
+
+    aws_byte_buf_secure_zero(&cc_key_pair->key_pair.key_buf);
+
+    struct aws_byte_cursor to_append = aws_byte_cursor_from_array(s_preamble, sizeof(s_preamble));
+    aws_byte_buf_append(&cc_key_pair->key_pair.key_buf, &to_append);
+
+    if (pub_x && pub_y) {
+        aws_byte_buf_append(&cc_key_pair->key_pair.key_buf, pub_x);
+        aws_byte_buf_append(&cc_key_pair->key_pair.key_buf, pub_y);
+    } else {
+        cc_key_pair->key_pair.key_buf.len += s_key_coordinate_size * 2;
+    }
+
+    if (priv_key) {
+        aws_byte_buf_append(&cc_key_pair->key_pair.key_buf, priv_key);
+    }
+
+    cc_key_pair->key_pair.impl = cc_key_pair;
+    cc_key_pair->key_pair.allocator = allocator;
+
+    if (pub_x) {
+        cc_key_pair->key_pair.pub_x.buffer = cc_key_pair->key_pair.key_buf.buffer + 1;
+        cc_key_pair->key_pair.pub_x.len = s_key_coordinate_size;
+
+        cc_key_pair->key_pair.pub_y.buffer = cc_key_pair->key_pair.pub_x.buffer + s_key_coordinate_size;
+        cc_key_pair->key_pair.pub_y.len = s_key_coordinate_size;
+    }
+
+    cc_key_pair->key_pair.priv_d.buffer = cc_key_pair->key_pair.key_buf.buffer + 1 + (s_key_coordinate_size * 2);
+    cc_key_pair->key_pair.priv_d.len = s_key_coordinate_size;
+    cc_key_pair->key_pair.vtable = &s_key_pair_vtable;
+
+    return cc_key_pair;
+
+error:
+    s_destroy_key_fn(&cc_key_pair->key_pair);
+    return NULL;
+}
+
+struct aws_ecc_key_pair *aws_ecc_key_pair_new_from_private_key(
+    struct aws_allocator *allocator,
+    enum aws_ecc_curve_name curve_name,
+    const struct aws_byte_cursor *priv_key) {
+    struct commoncrypto_ecc_key_pair *cc_key_pair =
+        s_alloc_pair_and_init_buffers(allocator, curve_name, NULL, NULL, priv_key);
+
+    if (!cc_key_pair) {
+        return NULL;
+    }
+
+    CFDataRef private_key_data = CFDataCreate(
+        cc_key_pair->cf_allocator, cc_key_pair->key_pair.key_buf.buffer, cc_key_pair->key_pair.key_buf.len);
+    CFMutableDictionaryRef key_attributes = CFDictionaryCreateMutable(cc_key_pair->cf_allocator, 6, NULL, NULL);
+    CFDictionaryAddValue(key_attributes, kSecAttrKeyType, kSecAttrKeyTypeECSECPrimeRandom);
+    CFDictionaryAddValue(key_attributes, kSecAttrKeyClass, kSecAttrKeyClassPrivate);
+    CFIndex key_size_bits = cc_key_pair->key_pair.priv_d.len * 8;
+    CFDictionaryAddValue(key_attributes, kSecAttrKeySizeInBits, &key_size_bits);
+    CFDictionaryAddValue(key_attributes, kSecAttrCanSign, kCFBooleanTrue);
+    CFDictionaryAddValue(key_attributes, kSecAttrCanVerify, kCFBooleanFalse);
+    CFDictionaryAddValue(key_attributes, kSecAttrCanDerive, kCFBooleanTrue);
```

Comment:
Not checking any of these return values

Reply:
AddValue doesn't have a return value, but I did add checks for everything else.